xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -248,6 +196,215 @@ private static ManagedChannel initLbChannel(
     return channel;
   }
 
+  /**
+   * A LoadBalancerProvider that provides a load balancer with a fixed XdsConfig.
+   */
+  private final class FixedXdsConfigBalancerProvider extends LoadBalancerProvider {
+    final XdsConfig xdsConfig;
+    final XdsConfig oldXdsConfig;
+    final EndpointWatcher oldEndpointWatcher;
+
+    FixedXdsConfigBalancerProvider(XdsConfig xdsConfig) {
+      this.xdsConfig = xdsConfig;
+      oldXdsConfig = LookasideLb.this.xdsConfig;
+      oldEndpointWatcher = endpointWatcher;
+    }
+
+    @Override
+    public boolean isAvailable() {
+      return true;
+    }
+
+    @Override
+    public int getPriority() {
+      return 5;
+    }
+
+    // A synthetic policy name identified by xds config.
+    @Override
+    public String getPolicyName() {
+      return "xds_policy__balancer_name_" + xdsConfig.balancerName
+          + "__childPolicy_" + xdsConfig.childPolicy
+          + "__fallbackPolicy_" + xdsConfig.fallbackPolicy
+          + "__edsServiceName_" + xdsConfig.edsServiceName
+          + "__lrsServerName_" + xdsConfig.lrsServerName;
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      return new FixedXdsConfigBalancer(helper);
+    }
+
+    final class FixedXdsConfigBalancer extends LoadBalancer {
+      final Helper helper;
+
+      // All fields become non-null once handleResolvedAddresses() successfully.
+      // All fields are assigned at most once.
+      @Nullable
+      ObjectPool<XdsClient> xdsClientRef;
+      @Nullable
+      XdsClient xdsClient;
+      @Nullable
+      LocalityStore localityStore;
+      @Nullable
+      LoadReportClient lrsClient;
+
+      FixedXdsConfigBalancer(Helper helper) {
+        this.helper = helper;
+      }
+
+      @Override
+      public void handleNameResolutionError(Status error) {}
+
+      @Override
+      public boolean canHandleEmptyAddressListFromNameResolution() {
+        return true;
+      }
+
+      @Override
+      public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        if (xdsClientRef != null) {
```

Comment:
Making this no-op is confusing. You should make this method not called if config is not changed. Then nothing needs to be added here. Neither bypassing silently nor a state check, as this is something completely internal of your implementation, and the caller guarantees this method would not be called for more than once.

Reply:
Why should the caller guarantee this method would not be called for more than once? I think that is wrong. The caller should know nothing about how the delegate will handle resolved address. The caller only knows it should switch balancer if config is changed.

Reply:
Callee assumes each invocation of this method gives the same `XdsConfig`. That is, it requires a stronger precondition than the method being overridden (which means the specification is weaker)!! You broke the inheritance.

All of the awkwardness comes from design and ideas of "config-fixed load balancer".

Reply:
I don't think the precondition is strong, the class name already indicates it. The caller and callee have a clear contract. I don't see any awkwardness.

Reply:
> I don't think the precondition is strong, the class name already indicates it.

This method requires a stronger precondition than the base class (which is `LoadBalancer`), therefore, [weaker specification](https://ocw.mit.edu/ans7870/6/6.005/s16/classes/07-designing-specs/#stronger_vs_weaker_specs). That's why this design not appropriate.

Well, all I mentioned might be too pedagogic to be productive. Now that this is how you implement internally, I would go with things that make it work at least. (But things will be very easily broken repeatedly in the future and you will end with endless debugging and fix. Also, inappropriate things is likely to spread, both in implementation and tests). That's why I request to reconsider class design. But, ... Too many things...